compiler/optimizer/OMRSimplifierHandlers.cpp

```diff
@@ -137,32 +137,160 @@
                   node);                                                           \
       }
 
-/**
- * Binary identity or zero operation
- *
- * If the second child is a constant that represents an identity operation,
- * replace this node with the first child.
- *
- * If the second child is a constant that represents a zero operation,
- * replace this node with the second child.
- */
-#define BINARY_IDENTITY_OR_ZERO_OP(ConstType,Type,NullValue,ZeroValue)\
-   if(secondChild->getOpCode().isLoadConst())\
-      {\
-      ConstType value = secondChild->get##Type();\
-      if(value == NullValue)\
-         return s->replaceNodeWithChild(node, firstChild, s->_curTree, block);\
-      if(value == ZeroValue)\
-         {\
-         if(performTransformation(s->comp(), "%sFound op with iconst in node [" POINTER_PRINTF_FORMAT "]\n", s->optDetailString(),node))\
-            {\
-            s->anchorChildren(node, s->_curTree);\
-            s->prepareToReplaceNode(node, secondChild->getOpCodeValue());\
-            node->set##Type(value);\
-            return node;\
-            }\
-         }\
+struct BinaryOpSimplifierHelpers
+   {
+   static int8_t getNodeByteValue(TR::Node * node)
+      {
+      return node->getByte();
+      }
+
+   static int16_t getNodeShortValue(TR::Node * node)
+      {
+      return node->getShortInt();
+      }
+
+   static int32_t getNodeIntValue(TR::Node * node)
+      {
+      return node->getInt();
+      }
+
+   static int64_t getNodeLongValue(TR::Node * node)
+      {
+      return node->getLongInt();
+      }
+
+   static uint32_t getNodeFloatBits(TR::Node * node)
+      {
+      return node->getFloatBits();
+      }
+
+   static uint64_t getNodeDoubleBits(TR::Node * node)
+      {
+      return node->getDoubleBits();
+      }
+
+   static void setNodeByteValue(TR::Node * node, int8_t value)
+      {
+      node->setByte(value);
+      }
+
+   static void setNodeShortValue(TR::Node * node, int16_t value)
+      {
+      node->setShortInt(value);
+      }
+
+   static void setNodeIntValue(TR::Node * node, int32_t value)
+      {
+      node->setInt(value);
+      }
+
+   static void setNodeLongValue(TR::Node * node, int64_t value)
+      {
+      node->setLongInt(value);
+      }
+
+   static void setNodeFloatBits(TR::Node * node, uint32_t bits)
+      {
+      node->setFloatBits(bits);
+      }
+
+   static void setNodeDoubleBits(TR::Node * node, uint64_t bits)
+      {
+      node->setDoubleBits(bits);
+      }
+   };
+
+template <typename T>
+struct BinaryOpSimplifier
+   {
+   TR::Simplifier * s;
+   T (* getNodeValue)(TR::Node *);
+   void (* setNodeValue)(TR::Node *, T);
+
+   /**
+   * Binary identity operation
+   *
+   * If the second child is a constant that represents an identity operation,
+   * replace this node with the first child.
+   */
+   inline TR::Node * tryAndSimplifyIdentityOp(TR::Node * node, T nullValue)
```

Comment:
I know the parameter had this name in the original version, but may I suggest renaming `nullValue` to `identityValue` for clarity?